quickwit-indexing/src/source/kafka_source.rs

```diff
@@ -68,29 +64,118 @@ impl TypedSourceFactory for KafkaSourceFactory {
     type Params = KafkaSourceParams;
 
     async fn typed_create_source(
-        source_id: String,
+        ctx: Arc<SourceExecutionContext>,
         params: KafkaSourceParams,
         checkpoint: SourceCheckpoint,
     ) -> anyhow::Result<Self::Source> {
-        KafkaSource::try_new(source_id, params, checkpoint).await
+        KafkaSource::try_new(ctx, params, checkpoint).await
     }
 }
 
-struct RdKafkaContext;
+enum RebalanceEvent {
+    Starting,
+    Assignments(HashMap<i32, PartitionId>)
+}
+struct RdKafkaContext {
+    ctx: Arc<SourceExecutionContext>,
+    rebalance_events: mpsc::Sender<RebalanceEvent>
+}
 
 impl ClientContext for RdKafkaContext {}
 
 impl ConsumerContext for RdKafkaContext {
     fn pre_rebalance(&self, rebalance: &Rebalance) {
-        info!("Pre rebalance {:?}", rebalance);
-    }
+        let source_id = self.ctx.config.source_id.clone();
+
+        let mut assignments = HashMap::new();
+
+        // TODO: need to also handle Rebalance::Revoke
+        if let &Rebalance::Assign(tpl) = rebalance {
+            // The rebalance API is not defined to be async so this attempts a work around
+            // by explicitly blocking the thread handling the rebalance callback until the
+            // database action is complete
+            let handle = Handle::current();
+            let _guard = handle.enter();
+            let result = futures::executor::block_on(async {
+                // This is intended to invalidate the in progress split. Since partitions may be
+                // moving away, any in progress state will be invalid and a new split would need
+                // to be started from the most recently committed offsets.
+                // TODO: this can throw an error
+                let _result = self.rebalance_events.send(RebalanceEvent::Starting).await;
+
+                self.ctx
+                    .metastore
+                    .index_metadata(&self.ctx.index_id)
```

Comment:
If we fetch the checkpoint in the prebalance handler, I think we cannot really prevent pre-rebalance splits traveling in the pipeline from being published after the rebalance.

Right now the most likely thing that will happen is this split being correctly published, and the postrebaalance pipeline failing / restarting once all of the time.

The failure will in turn trigger a rebalance and the system may or may not stabilize.
 
Is there a plan here?

Reply:
Yeah, anything that kills in flight splits needs to do it without taking down the kafka connection, otherwise it runs the risk of a rebalance cycle which isn't viable.  

If that's not possible then the serial number approach I described in #1794 should be sufficient to guard this but at the expense of having to complete all in flight splits before they can be invalidated at publication.